synapse/storage/engines/postgres.py

```diff
@@ -54,6 +61,13 @@ def on_new_connection(self, db_conn):
 
         cursor.close()
 
+    @property
+    def can_native_upsert(self):
+        """
+        Can we use native UPSERTs? This requires PostgreSQL 9.5+.
+        """
+        return self._version >= 90500
```

Comment:
This is going to blow up if called before `on_new_connection`. This is probably unlikely, but it may be worth defaulting it to something like `0`. Or have a `_can_do_native_upsert` parameter set to `False` by default instead, and have `on_new_connection` update it when its called

Reply:
This won't ever be called before we're actually connected, since it's in transaction code?